Add declaration-colon-space

jeddy3 opened the pull request:
Just wanted to get this up to help with the discussion around #38 and #40 

Will comment in those issues though... hang on.

jeddy3 commented:
I've had a go at adapting this in to a combined `-space` rule.

There were a few things I wasn't sure about so I figured I'd push up what I had and ask. I hope that's OK? Please bare with me as I continue to familiarise myself with nodejs JavaScript code :)

Currently it works with one or both of the options e.g. something like `{ "before": "always" }` (i.e. enforce before but not after) and something like `{ "before": "never", "after": "always" }` (i.e. enforce both before and after). I'm not sure if this is right.

For example, if you have:

```css
body {
  color:pink;
}
```

and

```js
{ "before": "always", "after": "always" }
```

It'll produce two warnings. One for `before` and one for `after`. Is that right, or should it produce a single warning e.g. "a single space was expected before and after the colon"? I assume it should as I wasn't sure how to add `notOK` tests for when two warnings were produced (hence the lack of `notOK` tests for `{ "before": "always", "after": "always" }` and the like).

I'm happy with either feedback here or, if it's easier, waiting on @davidtheclark to refactor his rules and then I will adapt this rule to follow his lead. Whichever is easiest for you guys.

Thanks!

davidtheclark commented:
Sorry I won't be able to work on this for another day or so. At that point I plan to try to figure out `selector-combinator-space`, so that might help. I'll take a look at this PR then too.

jeddy3 commented:
> At that point I plan to try to figure out selector-combinator-space, so that might help. I'll take a look at this PR then too.

Good for me :)

davidtheclark commented on src/rules/declaration-colon-space/__tests__/index.js:
```diff
@@ -0,0 +1,148 @@
+    "a { color : pink; }",
+    "single space before declaration's colon"
+  )
+  tr.notOk(
```
Might want to throw in another positive test or two ensuring that no spacing or lack thereof *after* the colon will cause a warning.

jeddy3 replied:
Will do

davidtheclark commented on src/rules/declaration-colon-space/__tests__/index.js:
```diff
@@ -0,0 +1,148 @@
+    "a { color : pink; }",
+    "single space before and after declaration's colon"
+  )
+})
```
Probably should throw in a negative test? Or do you not think it's necessary here?

jeddy3 replied:
Will do.

davidtheclark commented:
Nice job! Now I realize that my tests are incomplete :( --- I need to test before/after *alone* like you did.

If you and @MoOx agree with https://github.com/stylelint/stylelint/issues/40#issuecomment-101475582 we'll probably want to refactor the rule itself. But either way this is a really good start.

MoOx commented:
Refactor all the things :)

jeddy3 commented:
I've updated `declaration-colon-space` to reflect @davidtheclark's work on `declaration-bang-space`. Thanks again for that BTW, the standard whitespace stuff was a breeze to work with! :)

Umm, I merged in master but now I'm thinking I shouldn't have done that... right?

I've also updated the README.md to reflect the new options.

Looks OK?

davidtheclark commented:
Rad.

MoOx commented:
👌🏼
